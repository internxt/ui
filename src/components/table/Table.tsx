import React from 'react';

interface TableProps {
  children: React.ReactNode;
  className?: string;
}

interface TableHeaderProps {
  children: React.ReactNode;
  className?: string;
}

interface TableRowProps {
  children: React.ReactNode;
  className?: string;
  onClick?: (e: unknown) => void;
}

interface TableCellProps {
  children: React.ReactNode;
  className?: string;
  isHeader?: boolean;
  onClick?: (e: unknown) => void;
}

export const Table: React.FC<TableProps> = ({
  children,
  className = 'min-w-full border border-gray-10 rounded-lg overflow-hidden',
}) => (
  <div className={className}>
    <table className="w-full">{children}</table>
  </div>
);

export const TableHeader: React.FC<TableHeaderProps> = ({
  children,
  className = 'bg-gray-1 border-b border-gray-10 text-gray-100 font-semibold',
}) => <thead className={className}>{children}</thead>;

export const TableRow: React.FC<TableRowProps> = ({
  children,
  className = 'hover:bg-gray-1 border-b border-gray-10 last:border-none text-sm',
  onClick,
}) => (
  <tr onClick={onClick} className={className}>
    {children}
  </tr>
);

export const TableCell: React.FC<TableCellProps> = ({ children, className = 'p-4', isHeader = false, onClick }) => {
  const Component = isHeader ? 'th' : 'td';
  return (
    <Component onClick={onClick} className={`${className} ${isHeader ? 'text-left font-medium' : ''}`}>
      {children}
    </Component>
  );
};
