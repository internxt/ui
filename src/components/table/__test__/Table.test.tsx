import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Table, TableHeader, TableRow, TableCell, TableBody } from '../Table';

describe('Table', () => {
  it('renders table with default className', () => {
    render(
      <Table>
        <TableHeader>
          <TableRow>
            <TableCell isHeader>Example 1</TableCell>
            <TableCell isHeader>Example 2</TableCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>Cell 1</TableCell>
            <TableCell>Cell 2</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    );

    const tableComponent = screen.getByRole('table');
    expect(tableComponent).toBeInTheDocument();
    expect(tableComponent).toHaveClass('w-full');
    expect(tableComponent).toMatchSnapshot();
  });
});

describe('TableHeader', () => {
  it('renders header with default className', () => {
    render(
      <Table>
        <TableHeader>
          <TableRow>
            <TableCell isHeader>Example 1</TableCell>
            <TableCell isHeader>Example 2</TableCell>
          </TableRow>
        </TableHeader>
      </Table>,
    );

    const headerComponent = screen.getByRole('rowgroup');
    expect(headerComponent).toBeInTheDocument();
    expect(headerComponent).toHaveClass('bg-gray-1 border-b border-gray-10 text-gray-100 font-semibold');
    expect(headerComponent).toMatchSnapshot();
  });

  it('renders header with custom className', () => {
    render(
      <Table>
        <TableHeader className="custom-header-class">
          <TableRow>
            <TableCell isHeader>Example</TableCell>
          </TableRow>
        </TableHeader>
      </Table>,
    );

    const headerComponent = screen.getByRole('rowgroup');
    expect(headerComponent).toHaveClass('custom-header-class');
  });
});

describe('TableRow', () => {
  it('renders row with default className', () => {
    render(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>Example</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    );

    const rowComponent = screen.getByRole('row');
    expect(rowComponent).toBeInTheDocument();
    expect(rowComponent).toHaveClass('hover:bg-gray-1 border-b border-gray-10 last:border-none text-sm');
    expect(rowComponent).toMatchSnapshot();
  });

  it('renders row with custom className', () => {
    render(
      <Table>
        <TableBody>
          <TableRow className="custom-row-class">
            <TableCell>Example</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    );

    const rowComponent = screen.getByRole('row');
    expect(rowComponent).toHaveClass('custom-row-class');
  });

  it('calls onClick when row is clicked', () => {
    const handleClick = vi.fn();
    render(
      <Table>
        <TableBody>
          <TableRow onClick={handleClick}>
            <TableCell>Example</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    );

    const row = screen.getByRole('row');
    fireEvent.click(row);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});

describe('TableCell', () => {
  it('renders cell with default className', () => {
    render(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>Example</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    );

    const cell = screen.getByRole('cell');
    expect(cell).toBeInTheDocument();
    expect(cell).toHaveClass('p-4');
    expect(cell).toMatchSnapshot();
  });

  it('renders header cell with custom className and isHeader true', () => {
    render(
      <Table>
        <TableHeader>
          <TableRow>
            <TableCell isHeader className="custom-header-cell">
              Example
            </TableCell>
          </TableRow>
        </TableHeader>
      </Table>,
    );

    const headerCell = screen.getByRole('columnheader');
    expect(headerCell).toHaveClass('custom-header-cell text-left font-medium');
  });

  it('renders regular cell with custom className', () => {
    render(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell className="custom-cell-class">Example</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    );

    const cell = screen.getByRole('cell');
    expect(cell).toHaveClass('custom-cell-class');
  });

  it('calls onClick when cell is clicked', () => {
    const handleClick = vi.fn();
    render(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell onClick={handleClick}>Example</TableCell>
          </TableRow>
        </TableBody>
      </Table>,
    );

    const cell = screen.getByRole('cell');
    fireEvent.click(cell);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
