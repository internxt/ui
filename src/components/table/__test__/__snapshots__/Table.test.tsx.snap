// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > renders table with default className 1`] = `
<table
  class="w-full"
>
  <thead
    class="bg-gray-1 border-b border-gray-10 text-gray-100 font-semibold"
  >
    <tr
      class="hover:bg-gray-1 border-b border-gray-10 last:border-none text-sm"
    >
      <th
        class="p-4 text-left font-medium"
      >
        Example 1
      </th>
      <th
        class="p-4 text-left font-medium"
      >
        Example 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="hover:bg-gray-1 border-b border-gray-10 last:border-none text-sm"
    >
      <td
        class="p-4 "
      >
        Cell 1
      </td>
      <td
        class="p-4 "
      >
        Cell 2
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableCell > renders cell with default className 1`] = `
<td
  class="p-4 "
>
  Example
</td>
`;

exports[`TableHeader > renders header with default className 1`] = `
<thead
  class="bg-gray-1 border-b border-gray-10 text-gray-100 font-semibold"
>
  <tr
    class="hover:bg-gray-1 border-b border-gray-10 last:border-none text-sm"
  >
    <th
      class="p-4 text-left font-medium"
    >
      Example 1
    </th>
    <th
      class="p-4 text-left font-medium"
    >
      Example 2
    </th>
  </tr>
</thead>
`;

exports[`TableRow > renders row with default className 1`] = `
<tr
  class="hover:bg-gray-1 border-b border-gray-10 last:border-none text-sm"
>
  <td
    class="p-4 "
  >
    Example
  </td>
</tr>
`;
